File: python/Q2013_detect_squares.py
#
# @lc app=leetcode id=2013 lang=python3
#
# [2013] Detect Squares
#

# @lc code=start
class DetectSquares:
    def __init__(self):
        self.points: dict[tuple[int, int], int] = dict()

    def add(self, point: List[int]) -> None:
        self.points[(point[0], point[1])] = self.points.get((point[0], point[1]), 0) + 1

    def count(self, point: List[int]) -> int:
        x: int = point[0]
        y: int = point[1]

        result: int = 0

        for (i, j), count in self.points.items():
            if x == i or y == j or abs(x - i) != abs(y - j):
                continue

            result += count * self.points.get((x, j), 0) * self.points.get((i, y), 0)

        return result


# Your DetectSquares object will be instantiated and called as such:
# obj = DetectSquares()
# obj.add(point)
# param_2 = obj.count(point)
# @lc code=end

File: rust/src/leetcode/Q2013_detect_squares.rs
/*
 * @lc app=leetcode id=2013 lang=rust
 *
 * [2013] Detect Squares
 */

// @lc code=start
use std::collections::HashMap;

struct DetectSquares {
    points: HashMap<(i32, i32), i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl DetectSquares {
    fn new() -> Self {
        Self {
            points: HashMap::new(),
        }
    }

    fn add(&mut self, point: Vec<i32>) {
        let key: (i32, i32) = (point[0], point[1]);
        *self.points.entry(key).or_insert(0) += 1;
    }

    fn count(&self, point: Vec<i32>) -> i32 {
        let x: i32 = point[0];
        let y: i32 = point[1];

        let mut result: i32 = 0;

        for (&(i, j), &count) in &self.points {
            if x == i || y == j || (x - i).abs() != (y - j).abs() {
                continue;
            }

            result += count
                * self.points.get(&(x, j)).unwrap_or(&0)
                * self.points.get(&(i, y)).unwrap_or(&0);
        }

        result
    }
}

// /**
//  * Your DetectSquares object will be instantiated and called as such:
//  * let obj = DetectSquares::new();
//  * obj.add(point);
//  * let ret_2: i32 = obj.count(point);
//  */
// @lc code=end
